File: 규리/PGS 12905.py
# 1
def solution(board):
    N, M = len(board), len(board[0])
    DP = [[0]*M for _ in range(N)]
    mx = 0
    
    for i in range(N):
        for j in range(M):
            if board[i][j] == 1:
                DP[i][j] = min(DP[i-1][j], DP[i][j-1], DP[i-1][j-1]) + 1
            mx = max(mx, DP[i][j])

    return mx**2


# 2
def solution(board):
    N, M = len(board), len(board[0])
    MN = min(N, M)
    visited = [[[False]*M for _ in range(N)] for _ in range(MN)]
    answer = 0
    
    def find(x, y, mn):
        nonlocal answer, visited
        cnt = 0
        
        for i in range(x, x+mn):
            for j in range(y, y+mn):
                if board[i][j] == 1:
                    cnt += 1
        
        if cnt == mn**2:
            answer = max(answer, cnt)
        else:
            for i in range(N):
                for j in range(M):
                    if i + mn-1 <= N and j + mn-1 <= M \
                    and mn > 1 and visited[mn-2][i][j] != True:
                        find(i, j, mn-1)
    
    for i in range(N):
        for j in range(M):
            if i + MN <= N and j + MN <= M and visited[MN-1][i][j] != True:
                find(i, j, MN)

    return answer

File: 준수/PGS_12905.java
class Solution
{
    static int height;
    static int width;

    int[][] prefix = new int[1001][1001];
    
    public int solution(int [][]board)
    {
        int answer = board[0][0];
        
        height = board.length;
        width = board[0].length;

        for (int i = 1; i < height; i++) {
            for (int j = 1; j < width; j++) {
                if (board[i][j] == 1 && board[i - 1][j] == 1 && board[i][j - 1] == 1 && board[i - 1][j - 1] == 1) {
                    int min_value = 1001;
                    
                    if (min_value > prefix[i][j]) {
                        min_value = prefix[i][j];
                    }
                    
                    if (min_value > prefix[i][j + 1]) {
                        min_value = prefix[i][j + 1];
                    }
                    
                    if (min_value > prefix[i + 1][j]) {
                        min_value = prefix[i + 1][j];
                    }
                    
                    prefix[i + 1][j + 1] = min_value + 1;
                }
            }   
        }
        
        for (int i = 1; i < height + 1; i++) {
            for (int j = 1; j < width + 1; j++) {
                if (prefix[i][j] > 0 && prefix[i][j] + 1 > answer) answer = prefix[i][j] + 1;
            }
        }
        
        if (answer == 0 || answer == 1) {
            return answer;
        } else {
            return answer * answer;
        }
    }
}